zkstore/store.go

```diff
@@ -0,0 +1,446 @@
+func (s *Store) Delete(ident Ident) (found bool, err error) {
+	if err = ident.Validate(); err != nil {
+		return false, err
+	}
+	if ident.Version != "" {
+		return s.deleteVersion(ident)
+	}
+	return s.deleteItem(ident)
+}
+
+// deleteItem deletes the item and all versions within it
+func (s *Store) deleteItem(ident Ident) (found bool, err error) {
+	err = func() error {
+		var versions []string
+		versions, found, err = s.Versions(ident.Location)
+		switch {
+		case err != nil:
+			return err
+		case !found:
+			// the item could not be found. leave found=false and return
+			return nil
+		}
+		// delete each version of the specified item
+		for _, version := range versions {
+			versionIdent := ident
+			versionIdent.Version = version
+			versionIdent.ZKVersion = nil // force delete it no matter the zk version
+			if _, err = s.deleteVersion(versionIdent); err != nil {
+				return err
+			}
+		}
+		// and then delete the actual parent node
+		identPath, err := s.identPath(ident)
+		if err != nil {
+			return err
+		}
+		err = s.conn.Delete(identPath, ident.actualZKVersion())
+		switch err {
+		case zk.ErrNoNode:
+			return nil
+		case zk.ErrBadVersion:
+			return ErrVersionConflict
+		}
+		return err
+	}()
+	return found, err
+}
+
+// deleteVersion deletes only an item version
+func (s *Store) deleteVersion(ident Ident) (found bool, err error) {
+	err = func() error {
+		identPath, err := s.identPath(ident)
+		if err != nil {
+			return err
+		}
+		err = s.conn.Delete(identPath, ident.actualZKVersion())
+		switch err {
+		case zk.ErrNoNode:
+			// perhaps someone already deleted it? leave found==false
+			return nil
+		case zk.ErrBadVersion:
+			return ErrVersionConflict
+		}
+		found = true
+		return err
+	}()
+	return found, err
+}
+
+// List lists all of the known, latest version Locations that exist under
+// the specified category.  The locations will be sorted by Location Name.
+// If the category could not be found, found will be set to false.
+func (s *Store) List(category string) (locations []Location, found bool, err error) {
+	err = func() error {
+		if err = validateCategory(category); err != nil {
+			return errors.Wrap(err, "invalid category")
+		}
+		bucketsPath, err := s.bucketsPath(category)
+		if err != nil {
+			return err
+		}
+		buckets, _, err := s.conn.Children(bucketsPath)
+		switch {
+		case err == zk.ErrNoNode:
+			return nil
+		case err != nil:
+			return err
+		}
+		found = true
+		for _, bucket := range buckets {
+			children, _, err := s.conn.Children(path.Join(bucketsPath, bucket))
+			switch {
+			case err == zk.ErrNoNode:
+				// someone else deleted it? keep going.
+				continue
+			case err != nil:
+				return err
+			}
+			for _, child := range children {
+				pieces := strings.Split(child, "/")
+				locations = append(locations, Location{
+					Category: category,
+					Name:     pieces[len(pieces)-1],
+				})
+			}
+		}
+		return nil
+	}()
+	sort.Slice(locations, func(i, j int) bool {
```

Comment:
why sort? maybe let the caller sort if they really care? perhaps the API could provide a `ByName` func that clients could leverage?